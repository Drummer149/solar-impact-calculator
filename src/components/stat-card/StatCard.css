.statCard{
  flex: 1;
  height: 100%;
  width: 100%;
  display: flex;
  flex-flow: row;
  color: #fff;
  font-size: 3vmin;
  font-weight: 100;
  background-color: #05AC72;
}

.statCard svg {
  fill: #056644;
}

.statCard:hover {
  transform: scale(1.01);
  box-shadow: 0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23);
  z-index: 10;
}

.statCard:hover svg {
  animation: wiggle 0.82s cubic-bezier(.36,.07,.19,.97) both;
}

.statHalf {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
}

.text{
  background-color: #056644;
}

@keyframes wiggle {
  0% {transform: rotate(10deg);}
  25% {transform: rotate(-10deg);}
  50% {transform: rotate(20deg);}
  75% {transform: rotate(-5deg);}
  100% {transform: rotate(0deg);}
}

.odometer.odometer-auto-theme, .odometer.odometer-theme-minimal {
  display: inline-block;
  vertical-align: middle;
  *vertical-align: auto;
  *zoom: 1;
  *display: inline;
  position: relative;
}
.odometer.odometer-auto-theme .odometer-digit, .odometer.odometer-theme-minimal .odometer-digit {
  display: inline-block;
  vertical-align: middle;
  *vertical-align: auto;
  *zoom: 1;
  *display: inline;
  position: relative;
}
.odometer.odometer-auto-theme .odometer-digit .odometer-digit-spacer, .odometer.odometer-theme-minimal .odometer-digit .odometer-digit-spacer {
  display: inline-block;
  vertical-align: middle;
  *vertical-align: auto;
  *zoom: 1;
  *display: inline;
  visibility: hidden;
}
.odometer.odometer-auto-theme .odometer-digit .odometer-digit-inner, .odometer.odometer-theme-minimal .odometer-digit .odometer-digit-inner {
  text-align: left;
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow: hidden;
}
.odometer.odometer-auto-theme .odometer-digit .odometer-ribbon, .odometer.odometer-theme-minimal .odometer-digit .odometer-ribbon {
  display: block;
}
.odometer.odometer-auto-theme .odometer-digit .odometer-ribbon-inner, .odometer.odometer-theme-minimal .odometer-digit .odometer-ribbon-inner {
  display: block;
  -webkit-backface-visibility: hidden;
}
.odometer.odometer-auto-theme .odometer-digit .odometer-value, .odometer.odometer-theme-minimal .odometer-digit .odometer-value {
  display: block;
  -webkit-transform: translateZ(0);
}
.odometer.odometer-auto-theme .odometer-digit .odometer-value.odometer-last-value, .odometer.odometer-theme-minimal .odometer-digit .odometer-value.odometer-last-value {
  position: absolute;
}
.odometer.odometer-auto-theme.odometer-animating-up .odometer-ribbon-inner, .odometer.odometer-theme-minimal.odometer-animating-up .odometer-ribbon-inner {
  -webkit-transition: -webkit-transform 2s;
  -moz-transition: -moz-transform 2s;
  -ms-transition: -ms-transform 2s;
  -o-transition: -o-transform 2s;
  transition: transform 2s;
}
.odometer.odometer-auto-theme.odometer-animating-up.odometer-animating .odometer-ribbon-inner, .odometer.odometer-theme-minimal.odometer-animating-up.odometer-animating .odometer-ribbon-inner {
  -webkit-transform: translateY(-100%);
  -moz-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  -o-transform: translateY(-100%);
  transform: translateY(-100%);
}
.odometer.odometer-auto-theme.odometer-animating-down .odometer-ribbon-inner, .odometer.odometer-theme-minimal.odometer-animating-down .odometer-ribbon-inner {
  -webkit-transform: translateY(-100%);
  -moz-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  -o-transform: translateY(-100%);
  transform: translateY(-100%);
}
.odometer.odometer-auto-theme.odometer-animating-down.odometer-animating .odometer-ribbon-inner, .odometer.odometer-theme-minimal.odometer-animating-down.odometer-animating .odometer-ribbon-inner {
  -webkit-transition: -webkit-transform 2s;
  -moz-transition: -moz-transform 2s;
  -ms-transition: -ms-transform 2s;
  -o-transition: -o-transform 2s;
  transition: transform 2s;
  -webkit-transform: translateY(0);
  -moz-transform: translateY(0);
  -ms-transform: translateY(0);
  -o-transform: translateY(0);
  transform: translateY(0);
}