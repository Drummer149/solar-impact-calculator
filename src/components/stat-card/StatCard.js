import React, { Component } from 'react';
import Icon from '../icon/Icon';
import Moment from 'moment';
import Odometer from 'react-odometerjs';
import './StatCard.css';
import PropTypes from 'prop-types';

/**
 * Component to display a stat given some props
 */
class StatCard extends Component {

  constructor(props){
    super(props);
    this.state = {
      stat: 0,
      flowDirection: false,
      panelWattage: 310,
    }
  }

  // Find the amount of hours from a start to an enddate
  findHours(startDate, endDate){
    let duration = Moment.duration(startDate.diff(Moment(endDate)));
    return Math.round(duration.asHours());
  }

  // Convert hours and the number of solar panels to how many kilowatt hours the system produces
  convertHoursAndPanelsToKiloWattHours(hours, numPanels){
    return hours * 310 * numPanels / 2 / 1000;
  }

  // Given a kilowatthour and a multiplier return the statistic that will be displayed
  getStat(kph, multiplier){
    return parseInt(kph * multiplier, 10);
  }

  // Take in new props to set the statistic of the displaying card
  componentWillReceiveProps(props){
    let hours = this.findHours(Moment(), props.startDate);
    let kph = this.convertHoursAndPanelsToKiloWattHours(hours, props.currentInstallation);
    this.setState({stat : this.getStat(kph, props.metric.multiplier)});
  }

  // Should return css rule to reverse row if odd index
  getFlowDirection(index){
    return index % 2 === 1 ? 'row-reverse' : 'row';
  }
  
  render() {
    // Every second panel is reversed within flexbox
    let flowDirection = this.getFlowDirection(this.props.index);

    return (
      <div className="statCard" style={{ flexDirection : flowDirection }}>
        <div className="statHalf text" >
          <div style={{width: '80%'}}>
            YOU HAVE PREVENTED <Odometer value={this.state.stat} duration={1000 * this.props.index} format="d" /> {this.props.metric.title}
          </div>
        </div>
        <div className="statHalf logo">
          <Icon style={{fill: '#05AC72'}} name={this.props.metric.logo.toString()}/>
        </div>
      </div>
     )
   }
 }

 StatCard.propTypes = {
    index: PropTypes.number.isRequired,
    metric: PropTypes.object.isRequired,
    startDate: PropTypes.number.isRequired,
    currentInstallation: PropTypes.number.isRequired
 }

 export default StatCard;
 