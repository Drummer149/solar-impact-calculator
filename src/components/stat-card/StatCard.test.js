import React from 'react';
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import StatCard from './StatCard';
import Moment from 'moment';

Enzyme.configure({adapter: new Adapter()});

describe('<StatCard />', () => {
  const defaultProps = {
    index: 0, 
    key: 0, 
    startDate: Moment.now(),
    currentInstallation: 0,
    metric: 
      {
        title: 'METRIC TONNES OF CARBON DIOXIDE, ENTERING OUR ATMOSPHERE', 
        logo: 'carbon', 
        multiplier: 0.001
      }
    }

  test('Renders', () => {
    const wrapper = shallow(<StatCard {...defaultProps}/>);
    expect(wrapper.exists()).toBe(true);
  });

  test('Renders correctly', () => {
    const wrapper = shallow(<StatCard {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  })

  test('Has method get hours and will return the amount of hours between two dates', () => {
    const wrapper = shallow(<StatCard {...defaultProps}/>);
    expect(wrapper.instance().findHours(Moment('2019-01-01'), '2018-12-31')).toBe(24);
    expect(wrapper.instance().findHours(Moment('2019-01-01'), '2018-12-30')).toBe(48);
    expect(wrapper.instance().findHours(Moment('2019-01-01'), '2018-12-29')).toBe(72);
    expect(wrapper.instance().findHours(Moment('2019-01-01'), '2018-12-28')).toBe(96);
  });

  test('Has method convertHoursAndPanelsToKiloWattHours and will return the amount of kilowatt hours for particular solar system', () => {
    const wrapper = shallow(<StatCard {...defaultProps}/>);
    expect(wrapper.instance().convertHoursAndPanelsToKiloWattHours(24, 1)).toBe(3.72);
  });

  test('Has method getStat a will return the final statistic based on a kilowatt hour and a multiplier', () => {
    const wrapper = shallow(<StatCard {...defaultProps}/>);
    expect(wrapper.instance().getStat(300, 1)).toBe(300);
    expect(wrapper.instance().getStat(300, 0.5)).toBe(150);
    expect(wrapper.instance().getStat(300, 2)).toBe(600);
  });

  test('Has method getFlowDirection and will return css rule that will reverse flow direction on odd indexs', () => {
    const wrapper = shallow(<StatCard {...defaultProps}/>);
    expect(wrapper.instance().getFlowDirection(0)).toBe('row');
    expect(wrapper.instance().getFlowDirection(1)).toBe('row-reverse');
    expect(wrapper.instance().getFlowDirection(2)).toBe('row');
    expect(wrapper.instance().getFlowDirection(3)).toBe('row-reverse');
    expect(wrapper.instance().getFlowDirection(4)).toBe('row');
    expect(wrapper.instance().getFlowDirection(5)).toBe('row-reverse');
  });
});
