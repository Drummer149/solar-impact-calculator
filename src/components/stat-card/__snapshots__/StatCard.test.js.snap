// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatCard /> Renders correctly 1`] = `ShallowWrapper {}`;
